base_layer/core/src/base_node/validators/headers.rs

```diff
@@ -74,169 +76,19 @@ impl<B: BlockchainBackend> HeaderValidator<B> {
-                achieved,
-                target
-            );
-            return Err(ValidationError::BlockHeaderError(
-                BlockHeaderValidationError::ProofOfWorkError(PowError::AchievedDifficultyTooLow { achieved, target }),
-            ));
-        }
-
-        Ok(())
-    }
-
-    /// Returns the set of target difficulties for the given `BlockHeader`
-    fn fetch_target_difficulties(
-        &self,
-        block_header: &BlockHeader,
-    ) -> Result<Vec<(EpochTime, Difficulty)>, ValidationError>
-    {
-        let block_window = self
-            .rules
-            .consensus_constants(block_header.height)
-            .get_difficulty_block_window();
-        let start_height = block_header.height.saturating_sub(block_window);
-        if start_height == block_header.height {
-            return Ok(vec![]);
-        }
-
-        trace!(
-            target: LOG_TARGET,
-            "fetch_target_difficulties: new header height = {}, block window = {}",
-            block_header.height,
-            block_window
-        );
-
-        let block_window = block_window as usize;
-        // TODO: create custom iterator for chunks that does not require a large number of u64s to exist in memory
-        let heights = (0..block_header.height).rev().collect::<Vec<_>>();
-        let mut target_difficulties = Vec::with_capacity(block_window);
-        for block_nums in heights.chunks(block_window) {
-            let start = *block_nums.first().unwrap();
-            let end = *block_nums.last().unwrap();
-            let headers = self.db.fetch_headers(start, end)?;
-
-            let max_remaining = block_window.saturating_sub(target_difficulties.len());
-            trace!(
-                target: LOG_TARGET,
-                "fetch_target_difficulties: max_remaining = {}",
-                max_remaining
-            );
-            target_difficulties.extend(
-                headers
-                    .into_iter()
-                    .filter(|h| h.pow.pow_algo == block_header.pow.pow_algo)
-                    .take(max_remaining)
-                    .map(|h| (h.timestamp, h.pow.target_difficulty)),
-            );
-
-            assert!(
-                target_difficulties.len() <= block_window,
-                "target_difficulties can never contain more elements than the block window"
-            );
-            if target_difficulties.len() == block_window {
-                break;
-            }
-        }
-
-        trace!(
-            target: LOG_TARGET,
-            "fetch_target_difficulties: #returned = {}",
-            target_difficulties.len()
-        );
-        Ok(target_difficulties.into_iter().rev().collect())
+        // We cant use the actual tip, as this is used by the header sync, the tip has not yet been downloaded, but we
+        // can assume that the previous header was added, so we use that as the tip.
+        let tip_height = block_header.height.saturating_sub(1);
```

Comment:
The tip height passed in is incorrect. I suggest using the tip header.
```suggestion
             let tip_height = db.fetch_last_header()?.height;
```
From the usage in block validators:
```rust
     let tip_height = db.fetch_chain_metadata()?.height_of_longest_chain();
        check_median_timestamp(db, header, tip_height, self.rules.clone())?;
        trace!(
            target: LOG_TARGET,
            "BlockHeader validation: Median timestamp is ok for {} ",
            &header_id
        );
        check_achieved_and_target_difficulty(db, header, tip_height, self.rules.clone())?;
```

Reply:
I think this is actually fine

Reply:
This should be fine. Header sync is cant ask for tip header, as the meta data is only set after sync, to this will stay at 0. We need to ask the previous header and work on the assumption that it was added.